### PR TITLE
Laboratory work #2, Diana Gribanova - 22FPL1

### PR DESCRIPTION
**Pull request by Dzianna7**

### CODE REVIEW

**Review by marina-kaz — COMMENTED** [review]
Congratulations on passing lab 1!

There is nothing to review concerning lab 2.

**Comment by marina-kaz** [comment]
Closed as the student intends to continue working on the 1st lab.

File: lab_2_tokenize_by_bpe/main.py [hunk 1]
```diff
@@ -14,6 +14,16 @@ def prepare_word(
...
+    if not isinstance(raw_word, str) or not (isinstance(
+            start_of_word, str) or start_of_word is None) or not (
+            isinstance(end_of_word, str) or end_of_word is None):
+        return None
+    list_of_tokens = list(raw_word)
+    if start_of_word:
+        list_of_tokens.insert(0, start_of_word)
+    if end_of_word:
+        list_of_tokens.append(end_of_word)
+        return tuple(list_of_tokens)
```
Comment (marina-kaz) [inline lab_2_tokenize_by_bpe/main.py:26]: Is this the right indentation for `return`?

File: lab_2_tokenize_by_bpe/main.py [hunk 2]
```diff
@@ -26,6 +36,16 @@ def collect_frequencies(
...
     """
+    if not isinstance(text, str) or not isinstance(end_of_word, str) or not (
+            isinstance(start_of_word, str) or start_of_word is None):
+        return None
+    final_dictionary = {}
+    for i in text.split():
+        word = prepare_word(i, start_of_word, end_of_word)
+        if not word:
+            return None
+        final_dictionary.update({word: text.split().count(i)})
```
Comment (marina-kaz) [inline lab_2_tokenize_by_bpe/main.py:47]: `update` is suitable when you want to add multiple keys at a time. Here, it is better to add the key in a conventional way.

File: lab_2_tokenize_by_bpe/main.py [hunk 3]
```diff
@@ -36,6 +56,19 @@ def count_tokens_pairs(
...
+        return None
+    word_frequencies_list = list(word_frequencies)
+    print(word_frequencies_list)
+    dictionary = {}
+    for i in word_frequencies_list:
+        for j in range(len(i) - 1):
+            pair = (i[j], i[j + 1])
+            if pair not in dictionary:
+                dictionary[pair] = word_frequencies[i]
+            else:
```
Comment (marina-kaz) [inline lab_2_tokenize_by_bpe/main.py:69]: Consider rewriting it in such a way so that `else` is not needed

**Review by marina-kaz — COMMENTED** [review]
Good implementations!

File: lab_2_tokenize_by_bpe/main.py (hunk 1 above)
Comment (Dzianna7) [reply lab_2_tokenize_by_bpe/main.py:26]: done

File: lab_2_tokenize_by_bpe/main.py [hunk 4]
```diff
@@ -47,6 +79,19 @@ def merge_tokens(
     :param pair: a pair of tokens to be merged
     :return: dictionary in the form of <preprocessed word: number of occurrences>
     """
+    if not isinstance(word_frequencies, dict) or not isinstance(pair, tuple):
+        return None
+    dict_with_tokens = {}
+    pair_indexes = f"'{pair[0]}', '{pair[1]}'"
+    for elem in word_frequencies:
+        list_with_words = list(elem)
```
Comment (marina-kaz) [inline lab_2_tokenize_by_bpe/main.py:87]: Converting a list to a string is really, really creative... is this the best way?

File: lab_2_tokenize_by_bpe/main.py [hunk 5]
```diff
@@ -47,6 +79,19 @@ def merge_tokens(
...
+        if pair_indexes in str(elem):
+            for symbol in range(len(elem) - 1):
```
Comment (marina-kaz) [inline lab_2_tokenize_by_bpe/main.py:89]: This is a misleading variable name, it is not a symbol, right? It is the index of the symbol

File: lab_2_tokenize_by_bpe/main.py [hunk 6]
```diff
@@ -47,6 +79,19 @@ def merge_tokens(
...
+                if (elem[symbol] + elem[symbol + 1]) == pair[0] + pair[1]:
+                    list_with_words[symbol + 1] = ''.join(pair)
+                    list_with_words.pop(symbol)
+        dict_with_tokens.update({tuple(list_with_words): word_frequencies[elem]})
```
Comment (marina-kaz) [inline lab_2_tokenize_by_bpe/main.py:93]: Please don't use update to add a single key to the dictionary

**Review by marina-kaz — COMMENTED** [review]
Good progress! Keep going!

File: lab_2_tokenize_by_bpe/main.py (hunk 5 above)
Comment (Dzianna7) [reply lab_2_tokenize_by_bpe/main.py:89]: fixed

File: lab_2_tokenize_by_bpe/main.py (hunk 2 above)
Comment (Dzianna7) [reply lab_2_tokenize_by_bpe/main.py:47]: fixed

File: lab_2_tokenize_by_bpe/main.py (hunk 6 above)
Comment (Dzianna7) [reply lab_2_tokenize_by_bpe/main.py:93]: fixed

**Review by marina-kaz — COMMENTED** [review]
No changes

File: lab_2_tokenize_by_bpe/main.py (hunk 3 above)
Comment (Dzianna7) [reply lab_2_tokenize_by_bpe/main.py:68]: fixed

**Pull request closed without merging**